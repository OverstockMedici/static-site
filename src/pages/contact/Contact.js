import React, { useState, useEffect } from 'react';
import postData from './contact.post';
import { HeroImg, H2, Inquiries, MapImg, ContactH2, ContactForm, ServerMsg, Button } from './contact.style';
import { meta } from '../../content/contact.content';

const heroImgSm = '/img/coliseum-closeup-600.jpg';
const heroImgMed = '/img/coliseum-closeup-900.jpg';
const heroImgLg = '/img/coliseum-closeup-1200.jpg';
const mapImg = '/img/contact-map.png';
const mapImgSm = '/img/contact-map-600.jpg';


export default function Contact() {
    useEffect(() => {
        document.title = meta.title;
    }, []);

    const defaultState = {
        fname: '',
        lname: '',
        company: '',
        phone: '',
        email: '',
        subject: '',
        message: '',
        error: false,
        serverMsg: '',
        loading: false

    };
    const [state, setState] = useState(defaultState);
    const handleInputChange = (e) => {
        setState({
            ...state,
            [e.target.name]: e.target.value
        });
    };
    const handleSubmit = (e) => {
        e.preventDefault();
        setState({
            ...defaultState,
            loading: true
        });
        postData('/contact', state)
            .then(res => res.json())
            .then((response) => {
                setState({
                    ...defaultState,
                    loading: false,
                    serverMsg: response.message
                });
            })
            .catch((error) => {
                setState({
                    ...state,
                    error: true,
                    loading: false,
                    serverMsg: error.message || error
                });
            });
    };

    const getButtonContent = () => (
        state.loading ?
            <Button type="submit" disabled>Submitting</Button> :
            <Button type="submit">Submit</Button>
    );
    // TODO: viewStates
    // TODO: unit tests
    // TODO: replace fetch with axios or polyfill

    return (
        <div className="contact">
            <main role="main">
                <HeroImg>
                    <source media="(min-width: 1200px)" srcSet={heroImgLg} />
                    <source media="(min-width: 900px)" srcSet={heroImgMed} />
                    <source media="(max-width: 600px)" srcSet={heroImgSm} />
                    <img src={heroImgMed} alt="Peace Coliseum" />
                </HeroImg>

                <Inquiries>
                    <div className="inquiries-data">
                        <div className="inquery-info">
                            <H2>Business Inquries</H2>
                            <p>USA / Global</p>
                            <p>801.947.3100</p>
                            <p>info@mediciventures.com</p>
                        </div>
                        <div className="inquery-info">
                            <H2>PR Inquiries</H2>
                            <p>USA / Global</p>
                            <p>801.947.3564</p>
                            <p>pr@mediciventures.com</p>
                        </div>
                    </div>
                    <MapImg>
                        <source media="(min-width: 601px)" srcSet={mapImg} />
                        <source media="(max-width: 600px)" srcSet={mapImgSm} />
                        <img src={mapImgSm} alt="Medici Ventures Location" />
                    </MapImg>

                </Inquiries>
                <section className="contact-us">
                    <ContactH2>Contact Us</ContactH2>
                    {
                        state.serverMsg &&
                        <ServerMsg>{state.serverMsg}</ServerMsg>
                    }

                    {
                        !state.serverMsg &&
                        <ContactForm onSubmit={handleSubmit}>
                            <fieldset>
                                <legend>Name<sup>*</sup></legend>
                                <label className="sub-label">
                                    <input
                                        type="text"
                                        id="firstName"
                                        name="fname"
                                        value={state.fname}
                                        maxLength="30"
                                        required
                                        onChange={handleInputChange}
                                    />
                                    First Name
                                </label>
                                <label className="sub-label">
                                    <input
                                        type="text"
                                        id="lastName"
                                        name="lname"
                                        value={state.lname}
                                        maxLength="30"
                                        onChange={handleInputChange}
                                    />
                                    Last Name
                                </label>
                            </fieldset>
                            <label htmlFor="company">Company</label>
                            <input
                                type="text"
                                id="company"
                                name="company"
                                value={state.company}
                                maxLength="100"
                                onChange={handleInputChange}
                            />
                            <label htmlFor="email">Email Address<sup>*</sup></label>
                            <input
                                type="email"
                                id="email"
                                name="email"
                                value={state.email}
                                maxLength="100"
                                required
                                onChange={handleInputChange}
                            />
                            <label className="phone-number">
                                <input
                                    type="text"
                                    id="phone"
                                    name="phone"
                                    value={state.phone}
                                    maxLength="12"
                                    onChange={handleInputChange}
                                />
                            </label>
                            <label htmlFor="subject">Subject<sup>*</sup></label>
                            <input
                                type="text"
                                id="subject"
                                name="subject"
                                value={state.subject}
                                maxLength="100"
                                required
                                onChange={handleInputChange}
                            />
                            <label htmlFor="message">Message<sup>*</sup></label>
                            <textarea
                                id="message"
                                maxLength="500"
                                name="message"
                                value={state.message}
                                required
                                onChange={handleInputChange}
                            />
                            {getButtonContent()}
                        </ContactForm>
                    }
                </section>
            </main>
        </div>
    );
}
