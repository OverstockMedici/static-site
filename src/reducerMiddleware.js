const withDevTools = (
    process.env.NODE_ENV === 'development'
    && typeof window !== 'undefined'
    && window.__REDUX_DEVTOOLS_EXTENSION__
);
let devTools;

const reducerMiddleware = (reducer, initialState, name = 'React.context') => {
    if (withDevTools) {
        devTools = window.__REDUX_DEVTOOLS_EXTENSION__.connect({ name });
        devTools.init(initialState);
    }
    return (state, action) => {
        const newState = reducer(state, action);
        if (withDevTools) {
            devTools.send(action, newState);
        }
        return newState;
    };
};

export default reducerMiddleware;
