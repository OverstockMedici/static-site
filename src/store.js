import React from 'react';

export const initialState = { isMobile: true };

export const reducer = (state, action) => {
    switch (action.type) {
        case 'setDesktopLayout':
            return { ...state, isMobile: false };
        case 'setMobileLayout':
            return { ...state, isMobile: true };
        default:
            return state;
    }
};

export const Context = React.createContext();
