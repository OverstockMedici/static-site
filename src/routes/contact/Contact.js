import React, { useState, useContext } from 'react';
import Meta from '../../components/Meta';
import postData from './contact.post';
import {
    HeroImg,
    H2,
    Inquiries,
    MapImg,
    ContactH2,
    ContactForm,
    ServerMsg,
    Button,
    ErrorMessage,
} from './contact.style';
import { meta } from '../../content/contact.content';
import { Context } from '../../store';
import isEmail from './utils/isEmail';

const emailServiceUrl =
    'https://us-central1-corp-200921.cloudfunctions.net/medici-contact-form-service';

const heroImgSm = '/img/coliseum-closeup-600.jpg';
const heroImgMed = '/img/coliseum-closeup-900.jpg';
const heroImgLg = '/img/coliseum-closeup-1200.jpg';
const mapImg = '/img/contact-map.png';
const mapImgSm = '/img/contact-map-600.jpg';

export default function Contact() {
    const { store, dispatch } = useContext(Context);
    const defaultState = {
        error: false,
        formErrorMsg: '',
        serverMsg: '',
        loading: false,
    };
    const [state, setState] = useState(defaultState);
    const defaultFormData = {
        fname: '',
        lname: '',
        company: '',
        phone: '',
        email: '',
        subject: '',
        message: '',
    };
    const [formData, setFormData] = useState(defaultFormData);
    const handleInputChange = e => {
        setFormData({
            ...formData,
            [e.target.name]: e.target.value,
        });
    };

    const formValidation = formObject => {
        const isCorrectEmailFormat = isEmail(formObject.email);
        if (!isCorrectEmailFormat) {
            setState({
                formErrorMsg: 'Incorrect email format',
                loading: false,
            });
        }
    };

    const handleSubmit = e => {
        e.preventDefault();
        setState({
            ...defaultState,
            loading: true,
        });

        formValidation(formData);
        postData(emailServiceUrl, formData)
            .then(res => res.json())
            .then(response => {
                setState({
                    ...defaultState,
                    loading: false,
                    serverMsg: response.message,
                });
                setFormData({
                    ...defaultFormData,
                });
            })
            .catch(error => {
                setState({
                    ...state,
                    error: true,
                    loading: false,
                    serverMsg: error.message || error,
                });
            });
    };

    const getButtonContent = () =>
        state.loading ? (
            <Button type="submit" disabled isMobile={store.isMobile}>
                Submitting
            </Button>
        ) : (
            <Button type="submit" isMobile={store.isMobile}>
                Submit
            </Button>
        );
    // TODO: unit tests
    // TODO: replace fetch with axios or polyfill

    const { fname, lname, company, phone, email, subject, message } = formData;

    return (
        <div className="contact">
            <Meta {...meta} />
            <main role="main">
                <HeroImg>
                    <source media="(min-width: 1200px)" srcSet={heroImgLg} />
                    <source media="(min-width: 900px)" srcSet={heroImgMed} />
                    <source media="(max-width: 600px)" srcSet={heroImgSm} />
                    <img src={heroImgMed} alt="Peace Coliseum" />
                </HeroImg>

                <Inquiries>
                    <div className="inquiries-data">
                        <div className="inquery-info">
                            <H2>Business Inquries</H2>
                            <p>USA / Global</p>
                            <p>801.947.3100</p>
                            <p>info@mediciventures.com</p>
                        </div>
                        <div className="inquery-info">
                            <H2>PR Inquiries</H2>
                            <p>USA / Global</p>
                            <p>801.947.3564</p>
                            <p>pr@mediciventures.com</p>
                        </div>
                    </div>
                    <MapImg>
                        <source media="(min-width: 601px)" srcSet={mapImg} />
                        <source media="(max-width: 600px)" srcSet={mapImgSm} />
                        <img src={mapImgSm} alt="Medici Ventures Location" />
                    </MapImg>
                </Inquiries>
                <section className="contact-us">
                    <ContactH2>Contact Us</ContactH2>
                    {state.serverMsg && (
                        <ServerMsg>{state.serverMsg}</ServerMsg>
                    )}
                    {!state.serverMsg && (
                        <ContactForm onSubmit={handleSubmit}>
                            <fieldset>
                                <legend>
                                    Name<sup>*</sup>
                                </legend>
                                <label className="sub-label">
                                    <input
                                        type="text"
                                        id="firstName"
                                        name="fname"
                                        value={fname}
                                        maxLength="30"
                                        required
                                        onChange={handleInputChange}
                                    />
                                    First Name
                                </label>
                                <label className="sub-label">
                                    <input
                                        type="text"
                                        id="lastName"
                                        name="lname"
                                        value={lname}
                                        maxLength="30"
                                        onChange={handleInputChange}
                                    />
                                    Last Name
                                </label>
                            </fieldset>
                            <label htmlFor="company">Company</label>
                            <input
                                type="text"
                                id="company"
                                name="company"
                                value={company}
                                maxLength="100"
                                onChange={handleInputChange}
                            />
                            <label htmlFor="email">
                                Email Address<sup>*</sup>
                            </label>
                            <input
                                type="email"
                                id="email"
                                name="email"
                                value={email}
                                maxLength="100"
                                required
                                onChange={handleInputChange}
                            />
                            {state.formErrorMsg && (
                                <ErrorMessage>
                                    {state.formErrorMsg}
                                </ErrorMessage>
                            )}
                            <label className="phone-number">
                                <input
                                    type="text"
                                    id="phone"
                                    name="phone"
                                    value={phone}
                                    maxLength="12"
                                    onChange={handleInputChange}
                                />
                            </label>
                            <label htmlFor="subject">
                                Subject<sup>*</sup>
                            </label>
                            <input
                                type="text"
                                id="subject"
                                name="subject"
                                value={subject}
                                maxLength="100"
                                required
                                onChange={handleInputChange}
                            />
                            <label htmlFor="message">
                                Message<sup>*</sup>
                            </label>
                            <textarea
                                id="message"
                                maxLength="500"
                                name="message"
                                value={message}
                                required
                                onChange={handleInputChange}
                            />
                            {getButtonContent()}
                        </ContactForm>
                    )}
                </section>
            </main>
        </div>
    );
}
