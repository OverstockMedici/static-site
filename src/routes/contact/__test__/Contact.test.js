import React, { useReducer } from 'react';
import ReactDOM from 'react-dom';
import { Routes } from 'react-static';
import { Router, useLocation } from '@reach/router';

import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import { getByAltText } from '@testing-library/dom';

import { renderHook, act } from '@testing-library/react-hooks';

import { Context, reducer, initialState } from '../../../store';
import reducerMiddleware from '../../../reducerMiddleware';
import Contact from '../Contact';

afterEach(cleanup);

const reducerState =
    typeof window !== 'undefined' ? window?.__INITIAL_STATE__ : initialState;
const newReducer = reducerMiddleware(
    reducer,
    reducerState,
    'Medici Global Context',
);

const { result: reduceResult } = renderHook(() =>
    useReducer(newReducer, initialState),
);

// TODO: unit/integration tests are incomplete for this section
// I need more time on how to implement a mock request for emails

describe('Contact page logic', () => {
    const [store, dispatch] = reduceResult.current;

    const RenderContextAndRoutes = ({ children }) => (
        <Context.Provider value={{ store, dispatch }}>
            <Router primary={false}>{children}</Router>
        </Context.Provider>
    );

    test('Should failed with incorrect email format', async () => {
        const { getByTestId, getByText, debug } = render(
            <RenderContextAndRoutes>
                <Contact path="/" />
            </RenderContextAndRoutes>,
        );

        const submitButton = getByTestId('submit-button-mobile');
        const firstNameInput = getByTestId('first-name-input');
        const lastNameInput = getByTestId('last-name-input');
        const companyInput = getByTestId('company-input');
        const emailInput = getByTestId('email-input');
        const phoneInput = getByTestId('phone-input');
        const subjectInput = getByTestId('subject-input');
        const messageInput = getByTestId('message-input');

        fireEvent.change(firstNameInput, { target: { value: 'Liu' } });
        fireEvent.change(lastNameInput, { target: { value: 'Kang' } });
        fireEvent.change(companyInput, { target: { value: 'Earthrealm' } });
        fireEvent.change(emailInput, {
            target: { value: 'LiuKangThang@gmail' },
        });
        fireEvent.change(phoneInput, { target: { value: '911' } });
        fireEvent.change(subjectInput, { target: { value: 'Training' } });
        fireEvent.change(messageInput, { target: { value: 'Hi' } });

        fireEvent.click(submitButton);
        await waitFor(() => getByText('Incorrect email format'));
    });

    it('matches snapshot', () => {
        const contactTree = render(
            <RenderContextAndRoutes>
                <Contact path="/" />
            </RenderContextAndRoutes>,
        );

        expect(contactTree.firstChild).toMatchSnapshot();
    });
});
