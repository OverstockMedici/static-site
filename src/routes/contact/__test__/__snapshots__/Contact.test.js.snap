// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact page logic matches snapshot 1`] = `undefined`;
